### DISCUSSION TITLE
Metronome MIDI output not working

### DISCUSSION THREAD

**Discussion by newVideophile** [GrandOrgue/grandorgue#1997] [Help]
I am using GrandOrgue v3.15.0-1 with Friesach on a PC with LinuxMint.

The PC is connected to 2 keyboards and a pedalboard through USB. It is also connected to a Teensy 4.0, which I use for some custom controls. The Teensy can receive and send MIDI codes, for various purposes, such as replicating important information on a small alphanumeric LCD display, automatically sending registrations to GO, etc. 

Everything has been working well for several years so far, and the Teensy has been able to send and receive the MIDI codes I wanted. 

But recently, I wanted to duplicate the metronome information on the LCD display, i.e. on/off status, tempo and beats. To do so, the Teensy just needs to receive MIDI codes when any metronome control button is pressed. 

And here is the problem: whatever setting I use in GO, the only setting the Teensy can receive is the On/Off button. When I press the +1, -1, +10 and -10 buttons for the tempo, the Teensy receives nothing. Same for the +1 / -1 beat button. 

Here is the program I used on the Teensy for testing, I don't think I can make it simpler, nor do I see why it would not catch all MIDI events:

void setup() {
  Serial.begin(500000);
  while(!Serial);
}

void loop() {   
  if (usbMIDI.read()) {
    Serial.print(usbMIDI.getChannel());
    Serial.print("\t");
    Serial.print(usbMIDI.getData1());
    Serial.print("\t");
    Serial.println(usbMIDI.getData2());
  }
}

I know how to configure GO to send back MIDI codes, and, besides the metronome on/off button, I managed to do it successfully for stops activation/deactivation or key on/off events. The program above will then display the codes.

I tried various configurations, such as sending back controls, key on/off events, etc. but whatever the setting, the GrandOrgue MIDI output remains mute for metronome controls... 

What am I doing wrong?

[Edit] : GO actually sends the MIDI codes at startup, once for each metronome control (there is a total of 7 controls and I get 7 MIDI frames). And then never again. This doesn't really make sense to me...

**Comment by midimusic** [comment]
Your metronome buttons all need to be set up so that they output midi. Have you done this?
If not then right click the buttons and set them up. Use Midi-Ox to test the midi outputs (not your Teensy).
csw900

**Reply by newVideophile** [reply]
Yes I have set up the buttons to output midi. With the same setup, some buttons work, others don't.

I can't use Midi-ox, as it is not available for Linux. What Linux-compatible software would you recommend instead? gmidimonitor?

**Reply by midimusic** [reply]
I did realise that Midi-Ox was not a linux program after I has written my comment and closed down my computer. (It will run under Wine if you really need it.)

However you can test the GO's midi output with any other Linux based midi monitoring program. Until you have done this and reliably confirmed that GO is or is not outputting the expected midi there is no merit in any other discussion.

csw900

**Reply by newVideophile** [reply]
![GO](https://github.com/user-attachments/assets/25bf7148-eee2-43ba-85cd-0a5cbd15a89b)

So, I installed GMIDIMonitor as a replacement for Midi-ox: I got the same result as with the Teensy. I think we have a bug here...

**Comment by midimusic** [comment]
I have confirmed the results above with Windows and Midi-Ox. The ON button works but others do not. There is a bug.
Come in Oleg.

csw900

**Reply by willeke1234** [reply]
Do buttons send a MIDI message when they are clicked or when they turn on and off?

**Reply by newVideophile** [reply]
In my case, I don't click the buttons in the GO GUI. I have USB 61 key keyboard sending MIDI to GO. It has a few additional buttons that I have (successfully) used to control the GO metronome.

When I press a button on the keyboard, it sends a MIDI code to GO, which toogles the "On" button. Then, GO sends immediately back a code, with the last parameter being 0 (=off) or 127 (=on).

The other buttons are just controls, so I would expect them to send back a MIDI code each time they receive a MIDI command from the USB keyboard, i.e. at the moment I press down the corresponding button on the keyboard. 

Maybe the problem is not limited to the metronome, and GO handles properly the on/off buttons only. The stops, which are probably the most used ones, fall into this category

**Comment by willeke1234** [comment]
Can the Teensy extract the tempo and beats information from the SYSEX messages sent by the labels?

**Reply by newVideophile** [reply]
Yes, I tried that, but GO sends always the same data, regardless of the tempo and beats settings.
[Edit] : but it would be great if this could work too! Being able to directly get the value is safer than counting the number of +1, - 1, etc.

**Reply by willeke1234** [reply]
The SYSEX MIDI events from the labels work for me, although both labels are sent when something changes. Could you post the MIDI event settings and the MIDI monitor log?

**Reply by newVideophile** [reply]
I did a quick try this morning, but I lack the time to post the different things you are asking for. If I ask GO to display the MIDI events, then nothing is displayed, even if I set the SYSEX messages as an output for the tempo or beats. However, the Teensy receives the SYSEX messages, but always with the same content, as mentioned in my previous post. If the tempo is 80 and I press +1 twice, I would expect to get two different SYSEX sets of data, the first time coding 81, and the second time 82. This is not the case : the Teensy receives twice the very same series of bytes.

**Comment by newVideophile** [comment]
Is there a developer around?

**Comment by newVideophile** [comment]
Ok, I have found a workaround. The bug only affects buttons that don't switch between two states. This explains why the on/off button works, but the others don't (+1, -1 etc.). For those buttons, I send the MIDI control to other buttons that I never use (in the coupling panel), setting them to send back a MIDI control to the Teensy board (so that it gets the information about the setting, to be able to display it) and finally the Teensy sends back a MIDI commands to the metronom buttons. A bit complicated, but it works !
